File: scripts/python/src/tools/utils/sysinfo/bench/conditions.py
import os
import shutil
import subprocess
import sys

from tools.core.process import capture as run
from tools.utils.sysinfo.bench.capture import detect_virtualized
from tools.utils.sysinfo.bench.limits import (
    LOAD_LIMIT,
    MEMORY_FILESYSTEMS,
    MIN_FREE_DISK,
    THROTTLE_MARGIN,
)
from tools.utils.sysinfo.formatting import as_dict, as_list
from tools.utils.sysinfo.profiles import CPU_PROFILES, profile_for

GOVERNOR_PATH = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"

BATTERY_STATES = {"ac connected", "connected", "full", "fully charged"}


def cpu_governor():
    try:
        with open(GOVERNOR_PATH, encoding="utf-8") as handle:
            return handle.read().strip()
    except OSError:
        return ""


def on_battery(snapshot):
    batteries = as_list(snapshot.result("Battery", []))
    if not batteries:
        return False
    if as_list(snapshot.result("PowerAdapter", [])):
        return False
    for battery in batteries:
        status = str(battery.get("status") or "").lower()
        if status.startswith("charging") or status in BATTERY_STATES:
            return False
    return True


def load_average():
    try:
        return os.getloadavg()[0]
    except OSError:
        return None


def free_disk_ratio(path):
    try:
        stats = os.statvfs(path)
    except OSError:
        return None, None
    free = stats.f_bavail * stats.f_frsize
    total = stats.f_blocks * stats.f_frsize
    if not total:
        return free, None
    return free, free / total


def cpu_temperature(snapshot):
    cpu = as_dict(snapshot.result("CPU", {}))
    value = cpu.get("temperature")
    return value if isinstance(value, (int, float)) else None


def throttled(snapshot):
    temperature = cpu_temperature(snapshot)
    if temperature is None:
        return False
    cpu = as_dict(snapshot.result("CPU", {}))
    limit = profile_for(cpu.get("cpu") or "", CPU_PROFILES).max_temperature
    if not limit:
        return False
    return temperature >= limit - THROTTLE_MARGIN


THROTTLE_QUERIES = (
    "clocks_throttle_reasons.hw_thermal_slowdown",
    "clocks_throttle_reasons.sw_thermal_slowdown",
)


def nvidia_throttled():
    if not shutil.which("nvidia-smi"):
        return False
    # The timeout exists for a wedged driver, so the wedged driver must not then
    # surface as a traceback. capture_conditions runs before the job loop's
    # handler, and cool_down calls it again every 15s.
    try:
        result = run(
            ["nvidia-smi", f"--query-gpu={','.join(THROTTLE_QUERIES)}", "--format=csv,noheader"],
            timeout=3,
        )
    except (OSError, subprocess.SubprocessError):
        return False
    if result.returncode != 0:
        return False
    for line in result.stdout.strip().splitlines():
        for field in line.split(","):
            if field.strip().lower() == "active":
                return True
    return False


def capture_conditions(snapshot, workdir):
    free_bytes, free_ratio = free_disk_ratio(workdir)
    memory = as_dict(snapshot.result("Memory", {}))
    total = memory.get("total") or 0
    used = memory.get("used") or 0
    return {
        "on_battery": on_battery(snapshot),
        "governor": cpu_governor(),
        "loadavg_1": load_average(),
        "cpu_count": os.cpu_count(),
        "idle_temp_c": cpu_temperature(snapshot),
        "free_ram_bytes": max(total - used, 0) or None,
        "free_disk_bytes": free_bytes,
        "free_disk_ratio": free_ratio,
        "throttled_at_start": throttled(snapshot) or nvidia_throttled(),
        "virtualized": detect_virtualized(),
        "platform": sys.platform,
    }


def gate_reasons(conditions, writes_disk=False):
    reasons = []
    if conditions.get("on_battery"):
        reasons.append("running on battery")
    load = conditions.get("loadavg_1")
    count = conditions.get("cpu_count") or 1
    if load is not None and load / count > LOAD_LIMIT:
        reasons.append(f"load average is {load:.2f} across {count} logical cores")
    if conditions.get("throttled_at_start"):
        reasons.append("the machine is already thermally throttled")
    ratio = conditions.get("free_disk_ratio")
    if writes_disk and ratio is not None and ratio < MIN_FREE_DISK:
        reasons.append(f"only {ratio * 100:.0f}% of the filesystem is free")
    fstype = as_dict(conditions.get("filesystem")).get("fstype", "")
    if writes_disk and fstype in MEMORY_FILESYSTEMS:
        reasons.append(f"the work directory is on {fstype}, which measures memory and not disk")
    return tuple(reasons)


def grade_for(reasons, metrics, failures=()):
    if not metrics:
        return "aborted"
    # A run where suites crashed is degraded, not clean. Grading it clean made it
    # baseline-eligible and reset the staleness clock while measuring nothing.
    if reasons or failures:
        return "noisy"
    return "clean"

File: scripts/rust/crates/hwtune/src/bench/conditions.rs
use super::capture;
use serde_json::{Value, json};
use std::{fs, path::Path, process::Command};
use sysinfo::model::Snapshot;
pub fn on_battery(snapshot: &Snapshot) -> bool {
    let Some(batteries) = snapshot
        .result("Battery")
        .as_array()
        .filter(|b| !b.is_empty())
    else {
        return false;
    };
    if snapshot
        .result("PowerAdapter")
        .as_array()
        .is_some_and(|p| !p.is_empty())
    {
        return false;
    }
    !batteries.iter().any(|battery| {
        let status = sysinfo::formatting::text(&battery["status"]).to_lowercase();
        status.starts_with("charging")
            || ["ac connected", "connected", "full", "fully charged"].contains(&status.as_str())
    })
}
pub fn throttled(snapshot: &Snapshot) -> bool {
    let cpu = snapshot.result("CPU");
    cpu["temperature"]
        .as_f64()
        .zip(sysinfo::health::cpu_temperature_limit(
            cpu["cpu"].as_str().unwrap_or(""),
        ))
        .is_some_and(|(temperature, limit)| temperature >= limit - 5.0)
}
pub fn nvidia_throttled() -> bool {
    capture::probe(
        Command::new("nvidia-smi").args([
            "--query-gpu=clocks_throttle_reasons.hw_thermal_slowdown,clocks_throttle_reasons.sw_thermal_slowdown",
            "--format=csv,noheader",
        ]),
        3,
    ).is_some_and(|result| {
        result.status.success() && String::from_utf8_lossy(&result.stdout)
            .split([',', '\n']).any(|field| field.trim().eq_ignore_ascii_case("active"))
    })
}
pub fn capture_conditions(snapshot: &Snapshot, workdir: &Path) -> Value {
    let memory = snapshot.result("Memory");
    let total = memory["total"].as_u64().unwrap_or(0);
    let used = memory["used"].as_u64().unwrap_or(0);
    let free = fs2::available_space(workdir).ok();
    let disk_total = fs2::total_space(workdir).ok();
    json!({
        "on_battery": on_battery(snapshot),
        "governor": fs::read_to_string("/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor").unwrap_or_default().trim(),
        "loadavg_1": sysinfo_backend::System::load_average().one,
        "cpu_count": std::thread::available_parallelism().map_or(1, usize::from),
        "idle_temp_c": snapshot.result("CPU")["temperature"],
        "free_ram_bytes": total.saturating_sub(used),
        "free_disk_bytes": free,
        "free_disk_ratio": free.zip(disk_total).filter(|(_, total)| *total > 0).map(|(free, total)| free as f64 / total as f64),
        "throttled_at_start": throttled(snapshot) || nvidia_throttled(),
        "virtualized": capture::detect_virtualized(),
        "platform": capture::platform(),
    })
}
pub fn gate_reasons(conditions: &Value, writes_disk: bool) -> Vec<String> {
    let mut reasons = Vec::new();
    if conditions["on_battery"] == true {
        reasons.push("running on battery".into());
    }
    let count = conditions["cpu_count"]
        .as_f64()
        .filter(|n| *n > 0.0)
        .unwrap_or(1.0);
    if let Some(load) = conditions["loadavg_1"].as_f64()
        && load / count > 0.30
    {
        reasons.push(format!(
            "load average is {load:.2} across {count:.0} logical cores"
        ));
    }
    if conditions["throttled_at_start"] == true {
        reasons.push("the machine is already thermally throttled".into());
    }
    if writes_disk
        && let Some(ratio) = conditions["free_disk_ratio"].as_f64()
        && ratio < 0.20
    {
        reasons.push(format!(
            "only {:.0}% of the filesystem is free",
            ratio * 100.0
        ));
    }
    let fstype = conditions["filesystem"]["fstype"].as_str().unwrap_or("");
    if writes_disk && ["tmpfs", "ramfs", "devtmpfs", "hugetlbfs"].contains(&fstype) {
        reasons.push(format!(
            "the work directory is on {fstype}, which measures memory and not disk"
        ));
    }
    reasons
}
pub fn grade_for(reasons: &[String], metrics: usize, failures: &[String]) -> &'static str {
    if metrics == 0 {
        "aborted"
    } else if !reasons.is_empty() || !failures.is_empty() {
        "noisy"
    } else {
        "clean"
    }
}
